initial commit of R/Python virtual environments guide

erika-tyagi opened the pull request:
Initial draft of a short guide on using virtual environments for R (`renv`) and Python (`conda`). 

It's fairly opinionated, but (1) coming from a place of wanting to reduce the cognitive load of deciding which environment manager to use, and (2) favoring tools that have strong adoption/support within the broader data science and  reproducible research communities. 

I wanted to figure out how Quarto publishing to GitHub Pages works outside of the watchful eyes of UI-Research admins, but you can see a rendered version on my personal GitHub [here](https://erika-tyagi.github.io/virtual-environments/).

erika-tyagi committed:
initial commit

erika-tyagi committed:
update to use new urban quarto theme

lecy commented:
For the (1) why to use, I would add a more practical case: a new member joins a project team and needs to start running project scripts. Typically scripts include load commands but not install commands. These programs can help standardize project environments to ensure your computer is ready to run project code. 

Add a (2) section on WHEN (instead of why) you should use it? For example, if your project is advancing a specific methodology you want others to be able to replicate results. You might share data, scripts, plus a virtual environment so others can re-run the project code. Packages frequently change function names or arguments so code gets outdated quickly and without environment variables you need to de-bug code before it will run. 

I've learned this lesson when posting replication files for publications! https://lecy.github.io/political-ideology-of-nonprofits/

I get emails that the code is not working because I did not include a snapshot of the environment. 


I might add some details on the scope of a virtual environment. What are all of the pieces needed for reproducibility, and where does the virtual environment fit? For example - some other pieces to consider: 

![image](https://user-images.githubusercontent.com/1209099/184973634-44564361-fe1e-48da-9cb1-08c1570fc863.png)


Consider a section on how to break virtual environments - what are some things to consider that are outside the scope of `renv` or `conda` that could limit the usefulness of creating a virtual environment (rmarkdown depending on pandas, for example, which would not be indexed in the virtual environment). 

Another aspect of this is the mechanics of the virtual environments. They are capturing environmental settings in GitHub, but not archiving the actual code, correct? So this approach assumes the packages are retrievable from existing archives. Is that a safe assumption in most cases? Is there a work-around, for example adding a package RAR to GitHub so it can be installed locally if necessary?

erika-tyagi committed:
add to why and when sections, add xkcd comics

erika-tyagi commented:
> For the (1) why to use, I would add a more practical case: a new member joins a project team and needs to start running project scripts. Typically scripts include load commands but not install commands. These programs can help standardize project environments to ensure your computer is ready to run project code.

Good point, I added a bulleted list of concrete examples (all of which I've personally encountered)! 

> Add a (2) section on WHEN (instead of why) you should use it? For example, if your project is advancing a specific methodology you want others to be able to replicate results. You might share data, scripts, plus a virtual environment so others can re-run the project code. Packages frequently change function names or arguments so code gets outdated quickly and without environment variables you need to de-bug code before it will run.

Done, albeit I think the answer is that I think you should always set one up! 

> I've learned this lesson when posting replication files for publications! https://lecy.github.io/political-ideology-of-nonprofits/
> 
> I get emails that the code is not working because I did not include a snapshot of the environment.
> 
> I might add some details on the scope of a virtual environment. What are all of the pieces needed for reproducibility, and where does the virtual environment fit? For example - some other pieces to consider:
> 
> ![image](https://user-images.githubusercontent.com/1209099/184973634-44564361-fe1e-48da-9cb1-08c1570fc863.png)

Done, added a paragraph below the list of concrete examples (but as we talked about in Slack, I like the idea of keeping this guide discrete and well-scoped to save the broader context of reproducibility for another place). 

> Consider a section on how to break virtual environments - what are some things to consider that are outside the scope of `renv` or `conda` that could limit the usefulness of creating a virtual environment (rmarkdown depending on pandas, for example, which would not be indexed in the virtual environment).
> 
> Another aspect of this is the mechanics of the virtual environments. They are capturing environmental settings in GitHub, but not archiving the actual code, correct? So this approach assumes the packages are retrievable from existing archives. Is that a safe assumption in most cases? Is there a work-around, for example adding a package RAR to GitHub so it can be installed locally if necessary?

Not sure if I agree with this. I think there's a tradeoff between making a user guide comprehensive to cover every possible situation where things could go wrong vs. making it approachable to minimize the barrier to entry for new tools. For this guide, I'd love to err toward the latter (and let the official documentation take care of the former). I've also been using both `conda` and `renv` for a couple years now, and the steps here cover 99.9% of what I've ever had to do, so I don't want to risk making it seem like you need to have a deeper understanding or in-depth knowledge of more niche use cases to get utility out of adopting these tools. But I'm open to being overruled, as always!

erika-tyagi commented on index.qmd:
```diff
@@ -0,0 +1,234 @@
+
+To make this more concrete, here's a few situations when setting up a virtual environment can save you several hours (or even days!) of frustration: 
+
+* You published an analysis based on 2021 data. A year later, you want to update the analysis with 2022 data, but your code no longer works because a function in `dplyr` was updated. 
+* You need to update your version of `pandas` for a particular project, but a different project requires an older version of the library. 
+* Your coworker is running into errors running your code because of differences in package versions between your computers. 
+* You need to use Python 2.7 for a legacy project, while using Python 3.7 for all other projects. 
+* You want to use a new library that has known compatibility issues with Python 3.8 on Windows computers, so you want to use Python 3.7 for just one specific project. 
+* You got a new computer and don't want to spend hours remembering which R packages to install while resolving `Error in library(<>): there is no package called ‘<>’` errors. 
+* Your code takes a long time to run locally, so you want to leverage powerful [virtual computers in the cloud](http://tech-tools.urban.org/ec2-submission/) – or scale to use tens or hundreds or computers simultaneously – but you first need to install the relevant packages on the virtual machine(s). 
```
Very open to adding other examples (or cutting back if this seems too long), but I came up with this list based on personal experience/trauma.

lecy replied:
These are great!

erika-tyagi commented on index.qmd:
```diff
@@ -0,0 +1,234 @@
+
+## Getting help 
+
+Adding a new tool to your workflow can be hard, but we're here to help! Feel free to email Erika Tyagi (etyagi@urban.org) or drop a message in the #reproducible-research Slack channel with any questions or if you run into issues. 
```
I'd love to add at least one other person to this list – ideally someone with prior experience with both `conda` and `renv`?

lecy commented:
> Not sure if I agree with this.

All good - this was partly a question because I wasn't sure how often you would encounter the case where a package might not be retrievable (I did run into this case). But fundamentally I agree that the guide is meant to be instructive, not comprehensive, so including edge cases might be unproductive.

erika-tyagi committed:
fix typo and consolidate examples

erika-tyagi committed:
add note on mamba
